### PR TITLE
STYLE: Avoid array-like mutable default argument values

### PR DESCRIPTION
**Pull request by jhlegarreta**
- STYLE: Use `self` for the first argument to instance methods
- STYLE: Use `cls` for the first argument to class methods
- STYLE: Remove unused local variable
- DOC: Fix method parameter name in docstring
- DOC: Add missing method parameter to docstring
- DOC: Document optional arguments as such
- DOC: Fix typo in method documentation
- STYLE: Avoid array-like mutable default argument values

### CODE REVIEW

**Comment by jhlegarreta** [comment]
These fix a number of the many Python warnings.

File: dipy/core/graph.py [hunk 1]
```diff
@@ -63,7 +63,9 @@ def up_short(self, n):
     def down_short(self, n):
         return self.shortest_path(self.succ, n)
 
-    def all_paths(self, graph, start, end=None, path=[]):
+    def all_paths(self, graph, start, end=None, path=None):
+        if path is None:
+            path = []
```
Comment (skoudoro) [inline dipy/core/graph.py:68]: `path = path or []` instead

File: dipy/core/graph.py [hunk 2]
```diff
@@ -77,7 +79,9 @@ def all_paths(self, graph, start, end=None, path=[]):
                     paths.append(newpath)
         return paths
 
-    def shortest_path(self, graph, start, end=None, path=[]):
+    def shortest_path(self, graph, start, end=None, path=None):
+        if path is None:
+            path = []
```
Comment (skoudoro) [inline dipy/core/graph.py:84]: `path = path or []` instead

File: dipy/utils/parallel.py [hunk 3]
```diff
@@ -50,6 +50,13 @@ def paramap(func, in_list, out_shape=None, n_jobs=-1, engine="joblib",
     ndarray of identical shape to `arr`
 
     """
+
+    if func_args is None:
+        func_args = []
+
+    if func_kwargs is None:
+        func_kwargs = {}
```
Comment (skoudoro) [inline dipy/utils/parallel.py:58]: `func_args = func_args or []` instead

`func_kwargs = func_kwargs or {}` instead

File: dipy/workflows/reconst.py [hunk 4]
```diff
@@ -324,6 +324,9 @@ def run(self, input_files, bvalues_files, bvectors_files, mask_files,
            estimation of tensors by outlier rejection. MRM, 53: 1088-95.
 
         """
+        if save_metrics is None:
+            save_metrics = []
```
Comment (skoudoro) [inline dipy/workflows/reconst.py:328]: `save_metrics = save_metrics or []` instead

File: dipy/workflows/reconst.py [hunk 5]
```diff
@@ -815,6 +818,9 @@ def run(self, input_files, bvalues_files, bvectors_files, mask_files,
...
```
Comment (skoudoro) [inline dipy/workflows/reconst.py:822]: `save_metrics = save_metrics or []` instead

File: dipy/workflows/reconst.py [hunk 6]
```diff
@@ -979,6 +985,8 @@ def run(self, input_files, bvalues_files, bvectors_files, mask_files,
                        and perfusion in intravoxel incoherent motion MR
                        imaging." Radiology 168.2 (1988): 497-505.
         """
+        if save_metrics is None:
+            save_metrics = []
```
Comment (skoudoro) [inline dipy/workflows/reconst.py:989]: `save_metrics = save_metrics or []` instead

File: dipy/workflows/tests/test_reconst_dti.py [hunk 7]
```diff
@@ -23,7 +23,14 @@ def test_reconst_dti_alt_tensor():
...
 
 
-def reconst_flow_core(flow, extra_args=[], extra_kwargs={}):
+def reconst_flow_core(flow, extra_args=None, extra_kwargs=None):
+
+    if extra_args is None:
+        extra_args = []
+
+    if extra_kwargs is None:
+        extra_kwargs = {}
```
Comment (skoudoro) [inline dipy/workflows/tests/test_reconst_dti.py:32]: `func_args = func_args or []` instead

`func_kwargs = func_kwargs or {}` instead

**Review by skoudoro — COMMENTED** [review]
Hi @jhlegarreta,

Thank you for this! it is much appreciated. see below for some comments. After that, it should be ready to be merged

File: dipy/core/graph.py (hunk 1 above)
Comment (jhlegarreta) [reply dipy/core/graph.py]: Done in f49517e).

File: dipy/core/graph.py (hunk 2 above)
Comment (jhlegarreta) [reply dipy/core/graph.py]: Done in f49517e).

File: dipy/utils/parallel.py (hunk 3 above)
Comment (jhlegarreta) [reply dipy/utils/parallel.py]: Done in f49517e.

File: dipy/workflows/reconst.py (hunk 4 above)
Comment (jhlegarreta) [reply dipy/workflows/reconst.py]: Done in f49517e.

File: dipy/workflows/reconst.py (hunk 5 above)
Comment (jhlegarreta) [reply dipy/workflows/reconst.py]: Done in f49517e.

File: dipy/workflows/reconst.py (hunk 6 above)
Comment (jhlegarreta) [reply dipy/workflows/reconst.py]: Done in f49517e.

File: dipy/workflows/tests/test_reconst_dti.py (hunk 7 above)
Comment (jhlegarreta) [reply dipy/workflows/tests/test_reconst_dti.py]: Done in f49517e.

File: dipy/align/streamlinear.py [hunk 8]
```diff
@@ -877,6 +877,9 @@ def slr_with_qbx(static, moving,
     rm_small_clusters : int, optional
         Remove clusters that have less than `rm_small_clusters` (default 50)
 
+    maxiter : int
```
Comment (skoudoro) [inline dipy/align/streamlinear.py:880]: `int, optional` sorry, I forgot this one

Comment (jhlegarreta) [reply dipy/align/streamlinear.py]: Done in 409a7e0.

**Review by skoudoro — APPROVED** [review]
Thanks @jhlegarreta, all good, merging

**Pull request merged**